### PR TITLE
MRG : add geom.check for dipoles + update om_check_geom

### PR DESCRIPTION
**Pull request by agramfort**
aiming to fix https://neuroimage.usc.edu/forums/t/an-outlier-in-head-model-computed-by-openmeeg/11402/7

feedback welcome

still need a test...

I don't know if the om_assemble should crash in this case. Will it break the nested geometry case if I enforce this?

@eolivi @papadop any thoughts?

@massich can you see for the test? thanks

### CODE REVIEW

**Comment by agramfort** [comment]
@papadop @eolivi let me know if you're ok with this change.

cc @mclerc

**Comment by agramfort** [comment]
fascinating.... the new test passes here https://travis-ci.org/openmeeg/openmeeg/jobs/538821864 and fails here https://travis-ci.org/openmeeg/openmeeg/jobs/538821865

can someone replicate on his own machine?

**Comment by massich** [comment]
I can't replicate in my machine:
```sh
        Start 350: check_geom_dip
350/354 Test #350: check_geom_dip ....................   Passed    0.03 sec
```

**Comment by massich** [comment]
Rebased

**Comment by agramfort** [comment]
the 2 failures are related to this PR :(

**Comment by massich** [comment]
> the 2 failures are related to this PR :(

yes, I'm aware :D I'm just failing to make sense out of them. I'll keep pushing here

**Comment by massich** [comment]
With the following `DockerFile` I'm able to reproduce the error

```sh
# https://docs.travis-ci.com/user/common-build-problems/#Running-a-Container-Based-Docker-Image-Locally

# Using tag from https://hub.docker.com/r/travisci/ci-garnet/tags/ because tag 'latest' not found
# (see issue https://github.com/travis-ci/travis-ci/issues/7518)

# Environment analyze:
# * https://travis-ci.org/travis-ci-tester/travis-trusty-env/builds/279929194

FROM travisci/ci-garnet:packer-1515445631-7dfb2e1

MAINTAINER Joan Massich <mailsik@gmail.com>

USER travis

# From '.travis.yml':
# sources:
#   - ubuntu-toolchain-r-test
RUN sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
RUN sudo apt-get update

# From '.travis.yml':
# packages:
#   - libopenblas-dev
#   - liblapacke-dev
#   - swig
#   - python-numpy
#   - doxygen
#   - graphviz
RUN sudo apt-get -y install libopenblas-dev liblapacke-dev swig python-numpy doxygen graphviz

# From '.travis.yml':
# before_install:
RUN sudo dpkg --add-architecture i386
RUN sudo apt-get update
RUN sudo apt-get install -y libc6:i386 libncurses5:i386 libstdc++6:i386
RUN sudo apt-get install -y lcov
RUN sudo apt-get install binutils-2.26
ENV PATH="/usr/lib/binutils-2.26/bin:${PATH}"

# Install cmake v3.10.1
WORKDIR /home/travis/
RUN sudo apt-get purge cmake
RUN git clone https://github.com/astropy/ci-helpers.git
RUN /bin/bash ci-helpers/travis/setup_conda_linux.sh
ENV PATH="/home/travis/miniconda/bin:${PATH}"
RUN conda install cmake=3.12.2 --yes

# Setup Travis variables
WORKDIR /home/travis
ENV BACKEND=OpenBLAS
ENV USE_VTK=ON
ENV ENABLE_PYTHON=ON
ENV ANALYSE=ON
ENV ENABLE_COVERAGE=ON
ENV BUILD_DOCUMENTATION=ON

# Set up HOST variables to access local repo from the container
RUN mkdir -p /home/travis/code/openmeeg
RUN mkdir -p /home/travis/openmeeg_build/
WORKDIR /home/travis/openmeeg_build/
# ENV SRC_DIR=/home/travis/code/openmeeg/
# RUN cmake $SRC_DIR -DCMAKE_CXX_STANDARD=11 \
#                    -DBLA_VENDOR=$BACKEND \
#                    -DCMAKE_BUILD_TYPE=Debug \
#                    -DUSE_VTK=$USE_VTK \
#                    -DCMAKE_C_FLAGS_DEBUG="-g -O0" \
#                    -DCMAKE_CXX_FLAGS_DEBUG="-g -O0" 

# ADD /home/sik/code/openmeeg /home/travis/code/openmeeg
```

And this is what I get

```sh
travis@648a3d041b75:~/openmeeg_build$ ctest -VV -R check_geom_dip_outside
UpdateCTestConfiguration  from :/home/travis/openmeeg_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/travis/openmeeg_build/DartConfiguration.tcl
Test project /home/travis/openmeeg_build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 351
    Start 351: check_geom_dip_outside

351: Test command: /home/travis/openmeeg_build/apps/tools/om_check_geom "-g" "/home/travis/code/openmeeg/data/Head1/Head1.geom" "-d" "/home/travis/code/openmeeg/data/Head1/Head1_outside.dip"
351: Test timeout computed to be: 10000000
351: /home/travis/openmeeg_build/apps/tools/om_check_geom version 2.4.9999 compiled at Jul 15 2019 14:05:00 using OpenMP
351:  Executing using 4 threads.
351: 
351: This geometry is a NESTED geometry.
351: Info:: Mesh name/ID : 1
351: 		# vertices  : 42
351: 		# triangles : 80
351: 		Euler characteristic : 2
351: 		Min Area : 0.105436
351: 		Max Area : 0.125188
351: Info:: Mesh name/ID : 2
351: 		# vertices  : 42
351: 		# triangles : 80
351: 		Euler characteristic : 2
351: 		Min Area : 0.117904
351: 		Max Area : 0.139991
351: Info:: Mesh name/ID : 3
351: 		# vertices  : 42
351: 		# triangles : 80
351: 		Euler characteristic : 2
351: 		Min Area : 0.1393
351: 		Max Area : 0.165396
351: Info:: Domain name : Scalp
351: 		Conductivity : -1
351: 		Composed by interfaces : +2 -3 
351: 		Interface "2"= { mesh "2" }
351: 		Interface "3"= { mesh "3"(outermost) }
351: Info:: Domain name : Brain
351: 		Conductivity : -1
351: 		Composed by interfaces : -1 
351: 		Interface "1"= { mesh "1" }
351: Info:: Domain name : Air
351: 		Conductivity : -1
351: 		Composed by interfaces : +3 
351: 		Considered as the outermost domain.
351: 		Interface "3"= { mesh "3"(outermost) }
351: Info:: Domain name : Skull
351: 		Conductivity : -1
351: 		Composed by interfaces : +1 -2 
351: 		Interface "1"= { mesh "1" }
351: 		Interface "2"= { mesh "2" }
351: .geom : OK
351: .geom and .dip dipoles : OK
1/1 Test #351: check_geom_dip_outside ...........***Failed    0.02 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.05 sec

The following tests FAILED:
	351 - check_geom_dip_outside (Failed)
Errors while running CTest
```

File: OpenMEEG/src/geometry.cpp [hunk 1]
```diff
@@ -385,6 +385,26 @@ namespace OpenMEEG {
         return OK;
     }
 
+    bool Geometry::check(const Matrix& mat) const {
+        bool OK = true;
```
Comment (eolivi) [inline OpenMEEG/src/geometry.cpp]: We should check the size of `mat`, maybe assert there are at least 3 columns.

File: OpenMEEG/src/geometry.cpp [hunk 2]
```diff
@@ -385,6 +385,26 @@ namespace OpenMEEG {
...
+        if (!is_nested()) {
+            std::cerr << "Dipoles are only allowed when geometry is nested." << std::endl;
```
Comment (eolivi) [inline OpenMEEG/src/geometry.cpp]: IMO, we can have dipole anywhere. It should not be restricted to the inner_volume.
I remember some canadians who wanted to add dipoles in the scalp at eye location to model the eye blinks.
(but you could say `check_inner()` and later we'll extend this function)

Comment (agramfort) [reply OpenMEEG/src/geometry.cpp]: I like the idea of the check_inner. It's explicit.

@eolivi can you push directly to this PR if you have a bit of time?

**Comment by massich** [comment]
Ok. The problem is not with our code but with the fact that we are linking to different things:
```
libgomp.so.1
libstdc++.so.6
libgcc_s.so.1
libmatio.so.4
libhdf5.so.10
libz.so.1
```

If I pick them from conda it works, otherwise it doesn't.

**Comment by massich** [comment]
maybe we can use this as reference https://github.com/microsoft/GSL/blob/master/.travis.yml

File: OpenMEEG/src/geometry.cpp (hunk 1 above)
Comment (massich) [reply OpenMEEG/src/geometry.cpp]: this is taken Care by the `Matrix::operator()`

File: OpenMEEG/include/geometry.h [hunk 3]
```diff
@@ -95,6 +95,7 @@ namespace OpenMEEG {
         const bool&      is_nested()                      const { return is_nested_; }
               bool       selfCheck()                      const; ///< \brief the geometry meshes intersect each other
               bool       check(const Mesh& m)             const; ///< \brief check if m intersect geometry meshes
+              bool       check_inner(const Matrix& m)           const; ///< \brief check if dipoles are outside of geometry meshes
```
Comment (agramfort) [inline OpenMEEG/include/geometry.h]:
```suggestion
              bool       check_inner(const Matrix& m)     const; ///< \brief check if dipoles are outside of geometry meshes
```

**Comment by massich** [comment]
let's see what #379 says about appveyor

**Comment by agramfort** [comment]
ok for you @papadop @eolivi ?

@massich I suggest to ignore the appveyor failure as it's unrelated. We can fix it in the other PR

**Comment by massich** [comment]
sure.

**Comment by eolivi** [comment]
yes good for me too ! we can merge

**Comment by agramfort** [comment]
thx @eolivi for taking a look

**Pull request merged**